src/routes/api/auth/index.ts

```diff
@@ -24,22 +22,29 @@ const plugin: FastifyPluginAsyncTypebox = async (fastify) => {
     async function (request, reply) {
       const { username, password } = request.body
 
-      const user = await fastify.repository.find<Auth>('users', {
-        select: 'username, password',
-        where: { username }
-      })
+      const user = await fastify.knex<Credentials>('users')
+        .select('username', 'password')
+        .where({ username })
+        .first()
 
       if (user) {
         const isPasswordValid = await fastify.compare(password, user.password)
         if (isPasswordValid) {
-          const token = fastify.jwt.sign({ username: user.username })
+          const roles = await fastify.knex<{ name: string }>('roles')
+            .select('roles.name')
+            .join('user_roles', 'roles.id', '=', 'user_roles.role_id')
+            .join('users', 'user_roles.user_id', '=', 'users.id')
+            .where('users.username', username)
 
-          return { token }
+          request.session.user = { username, roles: roles.map(role => role.name) }
+
+          await request.session.save()
+
+          return { success: true }
```

Comment:
I think this should use a transaction

Reply:
```ts
async function (request, reply) {
      const { username, password } = request.body

      return fastify.knex.transaction(async (trx) => {
        try {
          const user = await fastify
            .knex<Credentials>('users')
            .select('username', 'password')
            .where({ username })
            .first()

          if (user) {
            const isPasswordValid = await fastify.compare(
              password,
              user.password
            )
            if (isPasswordValid) {
              const roles = await fastify
                .knex<{ name: string }>('roles')
                .select('roles.name')
                .join('user_roles', 'roles.id', '=', 'user_roles.role_id')
                .join('users', 'user_roles.user_id', '=', 'users.id')
                .where('users.username', username)

              request.session.user = {
                username,
                roles: roles.map((role) => role.name)
              }

              await request.session.save()

              return { success: true }
            }
          }

          reply.status(401)
          return { message: 'Invalid username or password.' }
        } catch (error) {
          await trx.rollback()
          throw error
        }
      })
    }
```

How would you test that transaction rollback on error?
I don't know if there is a way to mock/override decorators.

@climba03003

Reply:
I do see a way to monkey-patch `fastify.compare` and then throw at that point.

Reply:
I can do this:
```

test('Transaction should rollback on error', async (t) => {
  mock.module(
    path.join(
      import.meta.dirname,
      '../../../../src/plugins/custom/scrypt.js'
    ),
    {
      defaultExport: fastifyPlugin(
        async (fastify: FastifyInstance) => {
          fastify.decorate('compare', (value: string, hash: string) => {
            throw new Error()
          })
        },
        {
          name: 'scrypt'
        }
      )
    }
  )

  const app = await build(t)

  const res = await app.inject({
    method: 'POST',
    url: '/api/auth/login',
    payload: {
      username: 'basic',
      password: 'password123$'
    }
  })

  assert.strictEqual(res.statusCode, 500)
})
```

But it means using experimental mock.module feature and it weirdly breaks coverage...

Plz, can you give me the solution if you know it?

Reply:
![Scrypt-coverage](https://github.com/user-attachments/assets/241376a1-9c6e-4111-aec6-9747a41029ce)

Reply:
Ignore my previous messages, I don't know why I tough decorators were immutable...